overlord/hookstate/ctlcmd/stop.go

```diff
@@ -0,0 +1,132 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package ctlcmd
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/overlord/configstate"
+	"github.com/snapcore/snapd/overlord/hookstate"
+	"github.com/snapcore/snapd/overlord/servicectl"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+type stopCommand struct {
+	baseCommand
+	Positional struct {
+		ServiceNames []string `positional-arg-name:"<service>" required:"1"`
+	} `positional-args:"yes" required:"yes"`
+	Disable bool `long:"disable"`
+}
+
+var (
+	shortStopHelp = i18n.G("Stop services")
+)
+
+func init() {
+	addCommand("stop", shortStopHelp, "", func() command { return &stopCommand{} })
+}
+
+func getServiceInfos(st *state.State, snapName string, serviceNames []string) ([]*snap.AppInfo, error) {
```

Comment:
rather strange to have these helpers in `stop.go`. By “strange” I mean “arbitrary and confusing”. I don't mind having one file for each op (even though they're all fairly compact and really similar, meaning I myself would've probably plopped them all in a single file), but that means the generic bits need to live elsewhere.

Reply:
this function is very similar to `daemon.appInfosFor`, except that it's subtly different: in this version the arguments need to be `snap.service` (not `snap[.service]`), so you can't restart all the services with just `snap`. I think this negates the benefits of keeping things prefixed with the snap name. Why have it this way?

Reply:
If you're going to keep it in a separate func (and that might be fine, as this version can be much simpler than the other one which is more generic), why not make use of the fact that it's simpler? You've made it needlessly convoluted, I think.

Reply:
The generic version of ``daemon.appInfosFor`` is not easily reusable, unfortunately. It's generic, but also specific to the deamon (e.g. it's tied to http responses), and also it's it lives in the deamon which would create a circular dependency.
+1 for the comment re ``stop.go``, I'll move these functions to ``helpers.go`` or alike. The reason for having stop, start and restart separately is for consistency, as the existing snaptcl subcommands were implemented like that.

Reply:
I wasn't aware of the simplified syntax where you can only give snap name to stop all services so yes, need to handle that. Thanks for spotting.